### PR TITLE
When two targets match, a conversation is created

### PR DESCRIPTION
**Pull request by juan-apa**
When a match is created, a conversation is also created.
All the matches between two different users, are shared, e.g. when user_1 matches with user_2 a match and a conversation is created. If they match again with different targets, a match is created, but the conversation is reused for the new match.

### CODE REVIEW

File: app/models/match.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#  index_matches_on_target_creator_id     (target_creator_id)
+#
+
+class Match < ApplicationRecord
+  belongs_to :target_creator, class_name: 'Target'
+  belongs_to :target_compatible, class_name: 'Target'
+  belongs_to :conversation
+
+  before_validation :create_conversation
+  before_destroy :destroy_conversation
```
Comment (santir489) [inline app/models/match.rb]: I think this methods have to be private

File: app/models/message.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+class Message < ApplicationRecord
+  belongs_to :conversation
+  belongs_to :user
+
+  after_create :create_conversation
+
+  private
+
+  def create_conversation
+    self.conversation = Conversation.create!
```
Comment (santir489) [inline app/models/message.rb]: Is this `self` needed?

File: app/models/target.rb [hunk 3]
```diff
@@ -36,10 +37,16 @@ class Target < ApplicationRecord
...
+  before_destroy :delete_match
+
+  scope :target_matches, lambda { |id|
+    where(target_creator_id: id).or(Match.where(target_compatible_id: id)).pluck(:id)
+  }
 
   private
 
   def validate_target_limit
+    Rails.logger.info(matches.to_json)
```
Comment (santir489) [inline app/models/target.rb]: This has to do with `validate_target_limit`?

File: spec/models/target_spec.rb [hunk 4]
```diff
@@ -34,6 +34,104 @@
     }
   end
 
+  describe 'match creation/deletion' do
```
Comment (santir489) [inline spec/models/target_spec.rb]: I would use two different contexts

File: spec/models/target_spec.rb [hunk 5]
```diff
@@ -34,6 +34,104 @@
...
+        expect {
+          target_2.destroy!
+        }.to change { [Target.count, Match.count, Conversation.count] }
+          .from([2, 1, 1])
+          .to([1, 0, 0])
+      end
+    end
+
+    context 'creating a non-matching target' do
+      it 'doesn\'t create a match and a conversation' do
```
Comment (santir489) [inline spec/models/target_spec.rb]: I believe that `it "doesn't create a match and a conversation" do` is more readable

File: spec/models/target_spec.rb [hunk 6]
```diff
@@ -34,6 +34,104 @@
...
+      end
+      let!(:target_3) do
+        create :target,
+               latitude: target_1.latitude,
+               longitude: target_1.longitude,
+               topic: target_1.topic,
+               user: user_2
+      end
+
+      it 'deletes the target and the match, but keeps the conversation' do
```
Comment (santir489) [inline spec/models/target_spec.rb]: Maybe you could gruop context in a better way. I mean, you have
```
context 'deleting a matching target' do
..
end

context 'creating a non-matching target' do
...
end

context 'deleting a target that has multiple conversations with the same person' do
...
end

context 'creating a target that has already a match with the same user' do
...
end
```

File: app/models/conversation.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+# == Schema Information
+#
+# Table name: conversations
+#
+#  id :integer          not null, primary key
+#
+
+class Conversation < ApplicationRecord
+  has_many :matches, dependent: :delete_all
```
Comment (mdornel) [inline app/models/conversation.rb]: Are you sure that by destroying a conversation we want to delete the matches? If it's a requirement you can keep it, but I think it is unnecessary.

File: app/models/match.rb [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+#
+
+class Match < ApplicationRecord
+  belongs_to :target_creator, class_name: 'Target', inverse_of: :matches_creators
+  belongs_to :target_compatible, class_name: 'Target', inverse_of: :matches_compatible
+  belongs_to :user_creator, class_name: 'User', inverse_of: :matches_creator
+  belongs_to :user_compatible, class_name: 'User', inverse_of: :matches_compatible
+  belongs_to :conversation
+
+  before_validation :create_conversation
```
Comment (mdornel) [inline app/models/match.rb]: I don't think that `before_validation` is the right hook for this, take a look to [Object Lifecycle and available callbacks](https://guides.rubyonrails.org/active_record_callbacks.html#available-callbacks).

With `before_validation` you're going to trigger `create_conversation` method every time that the object is saved/updated. In addition to that `create_conversation` method looks really heavy (implement `joins`, `distinct`, etc) which is bad for performance if we keep it in a callback.

File: app/models/message.rb [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+#
+#  index_messages_on_conversation_id  (conversation_id)
+#  index_messages_on_user_id          (user_id)
+#
+
+class Message < ApplicationRecord
+  belongs_to :conversation
+  belongs_to :user
+
+  after_create :create_conversation
```
Comment (mdornel) [inline app/models/message.rb]: Do you really want to include this?

File: app/models/target.rb [hunk 10]
```diff
@@ -36,6 +46,11 @@ class Target < ApplicationRecord
   validate :validate_target_limit, on: :create
 
   after_create :create_matches
+  # before_destroy :delete_match
```
Comment (mdornel) [inline app/models/target.rb]: remove this

File: app/models/target.rb [hunk 11]
```diff
@@ -62,4 +83,9 @@ def matching_targets
           .where(topic_id: topic.id)
           .where.not(user_id: user.id)
   end
+
+  def delete_match
+    matches = Match.where(target_creator_id: id).or(Match.where(target_compatible_id: id))
+    matches.destroy_all
+  end
```
Comment (mdornel) [inline app/models/target.rb]: Remove this method, not necessary anymore.

File: app/models/target.rb [hunk 12]
```diff
@@ -36,6 +46,11 @@ class Target < ApplicationRecord
   validate :validate_target_limit, on: :create
 
   after_create :create_matches
+  # before_destroy :delete_match
+
+  scope :target_matches, lambda { |id|
+    where(target_creator_id: id).or(Match.where(target_compatible_id: id)).pluck(:id)
+  }
```
Comment (mdornel) [inline app/models/target.rb]: Is this scope necessary?

File: db/migrate/20191003174146_create_messages.rb [hunk 13]
```diff
@@ -0,0 +1,11 @@
+class CreateMessages < ActiveRecord::Migration[6.0]
+  def change
+    create_table :messages do |t|
+      t.belongs_to :user
+      t.belongs_to :conversation
+      t.string :body
```
Comment (mdornel) [inline db/migrate/20191003174146_create_messages.rb]: Maybe `text` type is more suitable, also I would add restrictions to this field, for example null constraints.

File: db/migrate/20191003174146_create_messages.rb [hunk 14]
```diff
@@ -0,0 +1,11 @@
+class CreateMessages < ActiveRecord::Migration[6.0]
+  def change
+    create_table :messages do |t|
+      t.belongs_to :user
+      t.belongs_to :conversation
+      t.text :body, { null: false }
```
Comment (santir489) [inline db/migrate/20191003174146_create_messages.rb]: I think braces are unnecessary here

**Pull request merged**